File: typescript/packages/core/src/core/dropbox/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DropboxAccessor } from '../../accessor/dropbox.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { dropboxDownload, dropboxDownloadStream } from './_client.ts'
import { readdir } from './readdir.ts'
import { rstripSlash, stripSlash } from '../../utils/slash.ts'
import { enoent } from '../../utils/errors.ts'

function eisdir(p: string): Error {
  const e = new Error(`EISDIR: ${p}`) as Error & { code: string }
  e.code = 'EISDIR'
  return e
}

function dropboxPathFromVirtual(virtualKey: string, prefix: string): string {
  let key = virtualKey
  if (prefix !== '' && key.startsWith(prefix)) key = key.slice(prefix.length)
  key = stripSlash(key)
  return key === '' ? '' : `/${key}`
}

export async function read(
  accessor: DropboxAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (key === '') throw eisdir(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`

  let entry = index !== undefined ? (await index.get(virtualKey)).entry : null
  if (entry === undefined || entry === null) {
    if (index !== undefined) {
      const parentKey = rstripSlash(virtualKey).replace(/\/[^/]+$/, '') || '/'
      if (parentKey !== virtualKey) {
        const parentPath = PathSpec.fromStrPath(parentKey, prefix)
        try {
          await readdir(accessor, parentPath, index)
          entry = (await index.get(virtualKey)).entry ?? null
        } catch {
          // parent refresh failed; fall through to ENOENT
        }
      }
    }
    if (entry === undefined || entry === null) throw enoent(path.original)
  }
  if (entry.resourceType === 'dropbox/folder') throw eisdir(path.original)
  const dropboxPath = dropboxPathFromVirtual(virtualKey, prefix)
  return dropboxDownload(accessor.tokenManager, dropboxPath)
}

export async function* stream(
  accessor: DropboxAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (key === '') throw eisdir(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`

  let entry = index !== undefined ? (await index.get(virtualKey)).entry : null
  if (entry === undefined || entry === null) {
    if (index !== undefined) {
      const parentKey = rstripSlash(virtualKey).replace(/\/[^/]+$/, '') || '/'
      if (parentKey !== virtualKey) {
        const parentPath = PathSpec.fromStrPath(parentKey, prefix)
        try {
          await readdir(accessor, parentPath, index)
          entry = (await index.get(virtualKey)).entry ?? null
        } catch {
          // parent refresh failed; fall through to ENOENT
        }
      }
    }
    if (entry === undefined || entry === null) throw enoent(path.original)
  }
  if (entry.resourceType === 'dropbox/folder') throw eisdir(path.original)
  const dropboxPath = dropboxPathFromVirtual(virtualKey, prefix)
  for await (const chunk of dropboxDownloadStream(accessor.tokenManager, dropboxPath)) {
    yield chunk
  }
}

export { dropboxPathFromVirtual }

File: python/mirage/core/hf_buckets/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import time

from opendal.exceptions import NotFound

from mirage.accessor.hf_buckets import HfBucketsAccessor
from mirage.cache.index import IndexCacheStore
from mirage.observe.context import record
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read_bytes(accessor: HfBucketsAccessor,
                     path: PathSpec,
                     index: IndexCacheStore | None = None,
                     offset: int = 0,
                     size: int | None = None) -> bytes:
    if isinstance(path, str):
        path = PathSpec.from_str_path(path)
    raw = path.strip_prefix
    key = raw.lstrip("/")
    op = accessor.operator()
    start_ms = int(time.monotonic() * 1000)
    try:
        if offset or size is not None:
            async with await op.open(key, "rb") as f:
                if offset:
                    await f.seek(offset)
                data = await f.read(size
                                    ) if size is not None else await f.read()
        else:
            data = bytes(await op.read(key))
    except NotFound as exc:
        raise enoent(path) from exc
    record("read", raw, accessor.RESOURCE_NAME, len(data), start_ms)
    return data
